cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -1911,5 +1911,50 @@ TEST(Cast, ExtensionTypeToIntDowncast) {
   }
 }
 
+TEST(Cast, DictTypeToAnotherDict) {
+  auto check_cast = [&](const std::shared_ptr<DataType>& in_type,
+                        const std::shared_ptr<DataType>& out_type,
+                        const std::string& json_str,
+                        const CastOptions& options = CastOptions()) {
+    auto arr = ArrayFromJSON(in_type, json_str);
+    auto exp = in_type->Equals(out_type) ? arr : ArrayFromJSON(out_type, json_str);
+    // this checks for scalars as well
+    CheckCast(arr, exp, options);
+  };
+
+  //    check same type passed on to casting
+  check_cast(dictionary(int8(), int16()), dictionary(int8(), int16()),
+             "[1, 2, 3, 1, null, 3]");
+  check_cast(dictionary(int8(), int16()), dictionary(int32(), int64()),
+             "[1, 2, 3, 1, null, 3]");
+  check_cast(dictionary(int32(), utf8()), dictionary(int8(), utf8()),
+             R"(["a", "b", "a", null])");
+
```

Comment:
Add tests casting from int to/from float and signed to/from unsigned integers.

Reply:
I tested
```
  check_cast(dictionary(int32(), int32()), dictionary(int8(), float32()),
             "[1, 2, 3, 1, null, 3]");
```
and get the following error:
`NotImplemented: JSON conversion to dictionary<values=float, indices=int8, ordered=0> not implemented`

Reply:
I ran int to/from float and unsigned (with Safe and Unsafe options) and it works correctly. LGTM.